### PR TITLE
esp32 RMT: Extend Micropython use of hardware carrier freq feature. 

### PR DESCRIPTION
**Pull request by JonRob812**
continuing from my premature pull request here: https://github.com/micropython/micropython/pull/6118

### CODE REVIEW

File: ports/esp32/esp32_rmt.c [hunk 1]
```diff
@@ -61,13 +64,25 @@ STATIC mp_obj_t esp32_rmt_make_new(const mp_obj_type_t *type, size_t n_args, siz
         { MP_QSTR_id,        MP_ARG_REQUIRED | MP_ARG_INT, {.u_int = -1} },
         { MP_QSTR_pin,       MP_ARG_REQUIRED | MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = mp_const_none} },
         { MP_QSTR_clock_div,                   MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = 8} }, // 100ns resolution
+        { MP_QSTR_carrier_duty_percent,        MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = 50} },
+        { MP_QSTR_carrier_freq_hz,             MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = MP_OBJ_NULL} },
```
Comment (dpgeorge) [inline ports/esp32/esp32_rmt.c:68]: following uses of frequency elsewhere, this doesn't need a `_hz` suffix (frequency is always in Hz in MicroPython); so it can just be `carrier_freq`

File: ports/esp32/esp32_rmt.c [hunk 2]
```diff
@@ -61,13 +64,25 @@ STATIC mp_obj_t esp32_rmt_make_new(const mp_obj_type_t *type, size_t n_args, siz
...
     };
     mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
     mp_arg_parse_all_kw_array(n_args, n_kw, all_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
     mp_uint_t channel_id = args[0].u_int;
     gpio_num_t pin_id = machine_pin_get_id(args[1].u_obj);
     mp_uint_t clock_div = args[2].u_int;
 
+    mp_obj_t carrier_en = mp_const_false;
```
Comment (dpgeorge) [inline ports/esp32/esp32_rmt.c:76]: this can just be a `bool`

File: ports/esp32/esp32_rmt.c [hunk 3]
```diff
@@ -103,8 +121,13 @@ STATIC mp_obj_t esp32_rmt_make_new(const mp_obj_type_t *type, size_t n_args, siz
...
     if (self->pin != -1) {
-        mp_printf(print, "RMT(channel=%u, pin=%u, source_freq=%u, clock_div=%u)",
-            self->channel_id, self->pin, APB_CLK_FREQ, self->clock_div);
+        if (self->carrier_en) {
+            mp_printf(print, "RMT(channel=%u, pin=%u, source_freq=%u, clock_div=%u, carrier_freq_hz=%u, carrier_duty_percent=%u)",
+                self->channel_id, self->pin, APB_CLK_FREQ, self->clock_div, self->carrier_freq_hz, self->carrier_duty_percent);
+        } else {
+            mp_printf(print, "RMT(channel=%u, pin=%u, source_freq=%u, clock_div=%u)",
+                self->channel_id, self->pin, APB_CLK_FREQ, self->clock_div);
+        }
```
Comment (dpgeorge) [inline ports/esp32/esp32_rmt.c:130]: to reduce duplication of code, better to keep the first part of the printing outside the if but without a closing `)`, then in the if just print the carrier parts

File: ports/esp32/esp32_rmt.c (hunk 1 above)
Comment (tve) [reply ports/esp32/esp32_rmt.c:68]: Why is this an obj, not an int?

Comment (dpgeorge) [reply ports/esp32/esp32_rmt.c:68]: It's so the code can detect between "not supplied" and "supplied".  If it's not supplied then the carrier freq is disabled.  Alternative would be to make -1 indicate disabled (and be the default value), and actually I think that would be a better way to do it so the user can explicitly select disabled by passing -1.

Comment (tve) [reply ports/esp32/esp32_rmt.c:68]: Or zero. A freq of zero is DC...

Comment (dpgeorge) [reply ports/esp32/esp32_rmt.c:68]: > Or zero. A freq of zero is DC...

Yes!

**Comment by dpgeorge** [comment]
@mattytrentini do you have any comments on this?

**Comment by mattytrentini** [comment]
I've actually [exposed similar functionality](https://github.com/micropython/micropython/compare/master...mattytrentini:esp32_rmt2) which I recently revisited in preparation for PR. Sorry @JonRob812, I should have submitted mine a while ago! But I think we can merge our efforts.

Obvious differences: @JonRob812 has added documentation. Mine adds the carrier level (which I thought was required but perhaps not). The bigger change on my branch is that I've also added different ways to specify the RMT pulse stream and I've exposed the idle level. 

I'll take a closer look tonight (and update the variable naming - I also used hz - that you suggested @dpgeorge).

File: ports/esp32/esp32_rmt.c (hunk 1 above)
Comment (JonRob812) [reply ports/esp32/esp32_rmt.c]: OK I changed so its enabled with an int greater than zero, default is zero

File: ports/esp32/esp32_rmt.c [hunk 4]
```diff
@@ -52,6 +52,9 @@ typedef struct _esp32_rmt_obj_t {
     uint8_t channel_id;
     gpio_num_t pin;
     uint8_t clock_div;
+    bool carrier_en;
```
Comment (dpgeorge) [inline ports/esp32/esp32_rmt.c:55]: this member can be removed

File: ports/esp32/esp32_rmt.c [hunk 5]
```diff
@@ -77,6 +92,9 @@ STATIC mp_obj_t esp32_rmt_make_new(const mp_obj_type_t *type, size_t n_args, siz
     self->channel_id = channel_id;
     self->pin = pin_id;
     self->clock_div = clock_div;
+    self->carrier_en = carrier_en;
```
Comment (dpgeorge) [inline ports/esp32/esp32_rmt.c:95]: no need to store this

File: ports/esp32/esp32_rmt.c [hunk 6]
```diff
@@ -85,11 +103,11 @@ STATIC mp_obj_t esp32_rmt_make_new(const mp_obj_type_t *type, size_t n_args, siz
     config.mem_block_num = 1;
     config.tx_config.loop_en = 0;
 
-    config.tx_config.carrier_en = 0;
+    config.tx_config.carrier_en = self->carrier_en;
```
Comment (dpgeorge) [inline ports/esp32/esp32_rmt.c:106]: just use `carrier_en`, not `self->carrier_en`

File: ports/esp32/esp32_rmt.c [hunk 7]
```diff
@@ -103,8 +121,14 @@ STATIC mp_obj_t esp32_rmt_make_new(const mp_obj_type_t *type, size_t n_args, siz
 STATIC void esp32_rmt_print(const mp_print_t *print, mp_obj_t self_in, mp_print_kind_t kind) {
     esp32_rmt_obj_t *self = MP_OBJ_TO_PTR(self_in);
     if (self->pin != -1) {
-        mp_printf(print, "RMT(channel=%u, pin=%u, source_freq=%u, clock_div=%u)",
+        mp_printf(print, "RMT(channel=%u, pin=%u, source_freq=%u, clock_div=%u",
             self->channel_id, self->pin, APB_CLK_FREQ, self->clock_div);
+        if (self->carrier_en) {
```
Comment (dpgeorge) [inline ports/esp32/esp32_rmt.c:126]: for this test use `self->carrier_freq > 0`

**Comment by dpgeorge** [comment]
> Obvious differences: @JonRob812 has added documentation. Mine adds the carrier level (which I thought was required but perhaps not).

@JonRob812 is there a reason you didn't include `carrier_level` as an option?  Maybe it's not needed after all, since all the IDF examples set it to 1.  Anyway, it can be easily added if needed (maybe later on?) by a new kwarg.

>  The bigger change on my branch is that I've also added different ways to specify the RMT pulse stream and I've exposed the idle level.

That sounds like a change that's better suited to a separate PR.

File: ports/esp32/esp32_rmt.c (hunk 7 above)
Comment (JonRob812) [reply ports/esp32/esp32_rmt.c:126]: > @JonRob812 is there a reason you didn't include `carrier_level` as an option? Maybe it's not needed after all, since all the IDF examples set it to 1. Anyway, it can be easily added if needed (maybe later on?) by a new kwarg.


@dpgeorge I see same thing looking at IDF docs. I would love to add it but it seems wrong to write code without fully understanding the context in which it operates.  Is it a simple invert? if so, it seems like this can be accomplished this with the `start=1` keyword in `write_pulses()`? 

Long story short, I lack of the full understanding that variable, `idle_output_en`, and `idle_level`.

**Comment by JonRob812** [comment]
> > Obvious differences: @JonRob812 has added documentation. Mine adds the carrier level (which I thought was required but perhaps not).
> 
> @JonRob812 is there a reason you didn't include `carrier_level` as an option? Maybe it's not needed after all, since all the IDF examples set it to 1. Anyway, it can be easily added if needed (maybe later on?) by a new kwarg.

I have no excuse other than I lacked understanding of how they get used. After researching I still lack the understanding. It seems like `carrier_level` does same thing that `start=1` vs `start=0` could do? I also dont understand `idle_output_en` or `idle_level`. What does idle mean? Is it sleep?

**Comment by mattytrentini** [comment]
If I understand correctly, when `idle_output_en` is enabled, the output of the pin will be `idle_level` before and after the RMT pulse sequence has completed. It's not clear to me what the output of the pin will be when `idle_output_en` is _not_ enabled but I would take a guess at it 'floating' - ie remaining at whatever level was drive by the last pulse.

`start` is not an IDF concept, it's used by the MicroPython RMT driver as an option to define the pulse stream. It could be used instead of `carrier_level` but I would prefer `carrier_level` to be exposed because `start` won't always be used to define the pulse stream (see my branch for alternative ways to configure it). Also, it's in the ESP RMT documentation so users would expect it. That said, it's not a commonly used feature so it's possibly not critical (but why not include it?).

**Comment by dpgeorge** [comment]
Thanks for applying the feedback.  As it stands I'd be happy to merge this PR.

Re `carrier_level`: I don't mind either way if this is added or not.  @mattytrentini maybe you want to follow up later on adding this and other enhancements?

**Comment by mattytrentini** [comment]
I'm ok if we add `carrier_level` later; it can wait.

I'm more interested in getting the different pulse stream definitions merged in; that seems to be more useful to folks...

**Comment by dpgeorge** [comment]
Squashed and merged in 1678f417445703750e187b1b8a83c8c3e0e3796f, with some typo fixes to the docs.

Thank you for the contribution!
